pkg/api/v2/reporter/reporter.go

```diff
@@ -155,21 +155,27 @@ func (r *reporter) WriteReports(ctx context.Context, resourceErrs ResourceReport
 			merr = multierror.Append(merr, err)
 			continue
 		}
-		resources.UpdateMetadata(resource, func(meta *core.Metadata) {
-			meta.ResourceVersion = updatedResource.GetMetadata().ResourceVersion
-		})
+		if updatedResource != nil {
+			resources.UpdateMetadata(resource, func(meta *core.Metadata) {
+				meta.ResourceVersion = updatedResource.GetMetadata().ResourceVersion
+			})
+		}
 
 		logger.Debugf("wrote report %v : %v", updatedResource.GetMetadata().Ref(), status)
```

Comment:
this also should be in the nil check. if the updated resource is nil, we didn't write a report. (and should log that we didn't write a report)

Reply:
we also should look into setting the resource that no longer exists to nil in `ResourceReports`, as we were side affecting those when we did successfully update the status.

this could be a dangerous change, need to look at the places we use `WriteReports` to make sure that it's safe

Reply:
I checked both the gloo and gateway syncers - both don't use the ResourceReports reports after it's passed in to the WriteReports call. When the new snapshot comes in and gets validated, which creates the new report, it should be updated to no longer have the resource that no longer exists.